cmd/kubeadm/app/phases/upgrade/prepull.go

```
@@ -0,0 +1,177 @@
+
+type prepuller interface {
+	CreateFunc(string)
+	WaitFunc(string)
+	DeleteFunc(string)
+}
+
+type DaemonSetPrepuller struct {
+	client clientset.Interface
+	cfg    *kubeadmapi.MasterConfiguration
+}
+
+func NewDaemonSetPrepuller(client clientset.Interface, cfg *kubeadmapi.MasterConfiguration) *DaemonSetPrepuller {
+	return &DaemonSetPrepuller{
+		client: client,
+		cfg:    cfg,
+	}
+}
+
+// CreateFunc creates a DaemonSet for making the image available on every relevant node
+func (d *DaemonSetPrepuller) CreateFunc(component string) {
+	ds := buildPrePullDaemonSet(d.cfg, component)
+
+	// Create the DaemonSet in the API Server
+	if _, err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Create(ds); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			fmt.Printf("[upgrade/prepull] WARNING: Failed to create daemonset for prepulling component %q: %v", component, err)
+			return
+		}
+
+		if _, err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Update(ds); err != nil {
+			// TODO: We should retry on 409 responses
+			fmt.Printf("[upgrade/prepull] WARNING: Failed to update daemonset for prepulling component %q: %v", component, err)
+		}
+	}
+}
+
+// WaitFunc waits for all Pods in the specified DaemonSet to be in the Running state
+func (d *DaemonSetPrepuller) WaitFunc(component string) {
+	fmt.Printf("[upgrade/prepull] Prepulling image for component %s.\n", component)
+	kubeadmutil.WaitForPodsWithLabel(d.client, 0, ioutil.Discard, "k8s-app=upgrade-prepull-"+component)
+}
+
+// DeleteFunc deletes the DaemonSet used for making the image available on every relevant node
+func (d *DaemonSetPrepuller) DeleteFunc(component string) {
+	dsName := addPrepullPrefix(component)
+	foregroundDelete := metav1.DeletePropagationForeground
+	deleteOptions := &metav1.DeleteOptions{
+		PropagationPolicy: &foregroundDelete,
+	}
+	if err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Delete(dsName, deleteOptions); err != nil {
```

should we return an error here? In other words, should a failed cleanup halt execution?